File: public/js/canonical.js
// Deterministic serialization for signed payloads. Must produce byte-identical
// output to lib/reputable_chat/crypto/canonical.rb or every signature fails;
// spec/canonical_parity_spec.rb checks that they agree.

function normalize(value) {
  if (Array.isArray(value)) return value.map(normalize);

  if (value && typeof value === "object") {
    const out = {};
    // Ruby sorts by byte order and JS by UTF-16 code unit. Every key we sign is
    // ASCII (fixed field names, or base64url public keys), where the two agree.
    for (const key of Object.keys(value).sort()) out[key] = normalize(value[key]);
    return out;
  }

  if (typeof value === "number" && !Number.isInteger(value)) {
    // Floats have no single textual form across languages. Anything signed
    // must arrive as a string or an integer.
    throw new Error(`refusing to canonicalize a float: ${value}`);
  }

  return value;
}

export function dump(object) {
  return JSON.stringify(normalize(object));
}

export function bytes(object) {
  return new TextEncoder().encode(dump(object));
}

File: lib/reputable_chat/crypto/canonical.rb
# frozen_string_literal: true

require "json"

module ReputableChat
  module Crypto
    # Deterministic serialization for anything that gets signed.
    #
    # The browser signs bytes and the server verifies bytes, so both sides have
    # to produce byte-identical output for the same object or every signature
    # fails. Keys are sorted, separators carry no whitespace, output is UTF-8.
    # This mirrors JSON.stringify over sorted keys in public/js/canonical.js --
    # the two must be changed together.
    module Canonical
      module_function

      def dump(object)
        JSON.generate(normalize(object))
      end

      def bytes(object)
        dump(object).b
      end

      def normalize(object)
        case object
        when Hash
          object.each_with_object({}) { |(k, v), out| out[k.to_s] = normalize(v) }
                .sort.to_h
        when Array   then object.map { |v| normalize(v) }
        when Symbol  then object.to_s
        when Float
          # Floats have no single textual form across languages. Anything that
          # needs to be signed must arrive as a string or an integer.
          raise ArgumentError, "refusing to canonicalize a Float: #{object}"
        else object
        end
      end
    end
  end
end
